File: Sources/CodexStatusCore/CodexStatusMonitor.swift
import Foundation

public enum CodexState: String, Equatable {
    case idle
    case thinking
}

public struct CodexStatusSnapshot: Equatable {
    public let state: CodexState
    public let activeSessionCount: Int
    public let latestEventAt: Date?
    public let latestEventType: String?
    public let latestSessionFile: URL?
    public let latestSessionTitle: String?
    public let activeSessions: [StatusSessionSummary]
    public let idleSessions: [StatusSessionSummary]
    public let activeSessionTitles: [String]
    public let idleSessionTitles: [String]
    public let scannedFileCount: Int
    public let staleAfter: TimeInterval
    public let codexHome: URL
    public let errorMessage: String?

    public var isThinking: Bool {
        state == .thinking
    }
}

public final class CodexStatusMonitor {
    public let codexHome: URL
    public let staleAfter: TimeInterval

    private let sessionsDirectory: URL
    private let fileManager: FileManager
    private let parser: SessionActivityParser
    private let scanLimit: Int
    private let titleIndexURL: URL

    public init(
        codexHome: URL = FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".codex"),
        staleAfter: TimeInterval = 30 * 60,
        scanLimit: Int = 100,
        fileManager: FileManager = .default
    ) {
        self.codexHome = codexHome
        self.staleAfter = staleAfter
        self.scanLimit = scanLimit
        self.fileManager = fileManager
        self.sessionsDirectory = codexHome.appendingPathComponent("sessions", isDirectory: true)
        self.parser = SessionActivityParser()
        self.titleIndexURL = codexHome.appendingPathComponent("session_index.jsonl")
    }

    public func snapshot(at now: Date = Date()) -> CodexStatusSnapshot {
        guard fileManager.fileExists(atPath: sessionsDirectory.path) else {
            return CodexStatusSnapshot(
                state: .idle,
                activeSessionCount: 0,
                latestEventAt: nil,
                latestEventType: nil,
                latestSessionFile: nil,
                latestSessionTitle: nil,
                activeSessions: [],
                idleSessions: [],
                activeSessionTitles: [],
                idleSessionTitles: [],
                scannedFileCount: 0,
                staleAfter: staleAfter,
                codexHome: codexHome,
                errorMessage: "没有找到 Codex sessions 目录"
            )
        }

        do {
            let files = try recentSessionFiles(limit: scanLimit)
            let titleIndex = SessionTitleIndex(fileURL: titleIndexURL).load()
            var latestActivity: SessionActivity?
            var activeActivities: [SessionActivity] = []
            var idleActivities: [SessionActivity] = []

            for file in files {
                guard var activity = try parser.parse(url: file.url, modifiedAt: file.modifiedAt) else {
                    continue
                }
                if let sessionID = sessionID(from: activity.fileURL) {
                    activity.title = titleIndex[sessionID] ?? activity.title
                }

                if isNewer(activity, than: latestActivity) {
                    latestActivity = activity
                }

                if activity.isOpenTask, isFresh(activity, now: now) {
                    activeActivities.append(activity)
                } else {
                    idleActivities.append(activity)
                }
            }

            let representative = newestActivity(in: activeActivities) ?? latestActivity
            let activeSessions = sessionSummaries(from: activeActivities)
            let idleSessions = sessionSummaries(from: idleActivities)

            return CodexStatusSnapshot(
                state: activeActivities.isEmpty ? .idle : .thinking,
                activeSessionCount: activeActivities.count,
                latestEventAt: representative?.latestEventAt,
                latestEventType: representative?.latestEventType,
                latestSessionFile: representative?.fileURL,
                latestSessionTitle: representative.map(sessionTitle),
                activeSessions: activeSessions,
                idleSessions: idleSessions,
                activeSessionTitles: activeSessions.map(\.title),
                idleSessionTitles: idleSessions.map(\.title),
                scannedFileCount: files.count,
                staleAfter: staleAfter,
                codexHome: codexHome,
                errorMessage: nil
            )
        } catch {
            return CodexStatusSnapshot(
                state: .idle,
                activeSessionCount: 0,
                latestEventAt: nil,
                latestEventType: nil,
                latestSessionFile: nil,
                latestSessionTitle: nil,
                activeSessions: [],
                idleSessions: [],
                activeSessionTitles: [],
                idleSessionTitles: [],
                scannedFileCount: 0,
                staleAfter: staleAfter,
                codexHome: codexHome,
                errorMessage: error.localizedDescription
            )
        }
    }

    private func recentSessionFiles(limit: Int) throws -> [SessionFile] {
        let resourceKeys: Set<URLResourceKey> = [.isRegularFileKey, .contentModificationDateKey]
        guard let enumerator = fileManager.enumerator(
            at: sessionsDirectory,
            includingPropertiesForKeys: Array(resourceKeys),
            options: [.skipsHiddenFiles, .skipsPackageDescendants]
        ) else {
            return []
        }

        var files: [SessionFile] = []
        for case let url as URL in enumerator where url.pathExtension == "jsonl" {
            let values = try? url.resourceValues(forKeys: resourceKeys)
            guard values?.isRegularFile == true, let modifiedAt = values?.contentModificationDate else {
                continue
            }
            files.append(SessionFile(url: url, modifiedAt: modifiedAt))
        }

        return Array(files.sorted { $0.modifiedAt > $1.modifiedAt }.prefix(limit))
    }

    private func isFresh(_ activity: SessionActivity, now: Date) -> Bool {
        let freshnessDate = max(activity.latestEventAt ?? activity.modifiedAt, activity.modifiedAt)
        return now.timeIntervalSince(freshnessDate) <= staleAfter
    }

    private func isNewer(_ activity: SessionActivity, than other: SessionActivity?) -> Bool {
        guard let other else {
            return true
        }
        return (activity.latestEventAt ?? activity.modifiedAt) > (other.latestEventAt ?? other.modifiedAt)
    }

    private func newestActivity(in activities: [SessionActivity]) -> SessionActivity? {
        activities.max { lhs, rhs in
            (lhs.latestEventAt ?? lhs.modifiedAt) < (rhs.latestEventAt ?? rhs.modifiedAt)
        }
    }

    private func sessionSummaries(from activities: [SessionActivity]) -> [StatusSessionSummary] {
        activities
            .sorted {
                ($0.latestEventAt ?? $0.modifiedAt) > ($1.latestEventAt ?? $1.modifiedAt)
            }
            .map { activity in
                StatusSessionSummary(
                    id: sessionIdentifier(for: activity),
                    title: sessionTitle(for: activity),
                    lastAnswer: activity.lastAnswer
                )
            }
    }

    private func sessionTitle(for activity: SessionActivity) -> String {
        SessionTitleNormalizer.displayTitle(activity.title)
    }

    private func sessionIdentifier(for activity: SessionActivity) -> String {
        sessionID(from: activity.fileURL) ?? activity.fileURL.path
    }

    private func sessionID(from url: URL) -> String? {
        let baseName = url.deletingPathExtension().lastPathComponent
        guard baseName.count >= 36 else {
            return nil
        }

        let suffix = String(baseName.suffix(36))
        let uuidPattern = #"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"#
        return suffix.range(of: uuidPattern, options: .regularExpression) == nil ? nil : suffix
    }
}

private struct SessionFile {
    let url: URL
    let modifiedAt: Date
}

private struct SessionTitleIndex {
    let fileURL: URL

    func load() -> [String: String] {
        guard let text = try? String(contentsOf: fileURL, encoding: .utf8) else {
            return [:]
        }

        var titlesByID: [String: String] = [:]
        for line in text.split(separator: "\n") {
            guard let data = String(line).data(using: .utf8),
                  let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let id = object["id"] as? String,
                  let title = SessionTitleNormalizer.explicitTitle(in: object)
            else {
                continue
            }

            titlesByID[id] = title
        }

        return titlesByID
    }
}

struct SessionActivity {
    let fileURL: URL
    let modifiedAt: Date
    var title: String?
    var latestEventAt: Date?
    var latestEventType: String?
    var lastTaskStartedAt: Date?
    var lastTaskCompletedAt: Date?
    var lastAnswer: String?
    var sawAnyEvent = false

    var isOpenTask: Bool {
        if let lastTaskStartedAt, let lastTaskCompletedAt {
            return lastTaskStartedAt > lastTaskCompletedAt
        }

        if lastTaskStartedAt != nil {
            return true
        }

        if lastTaskCompletedAt != nil {
            return false
        }

        // A completed Codex task writes task_complete at the end. If a recently
        // modified tail has activity but no completion marker, treat it as open.
        return sawAnyEvent
    }
}

final class SessionActivityParser {
    private let tailByteLimit: UInt64
    private let isoWithFractionalSeconds: ISO8601DateFormatter
    private let isoWithoutFractionalSeconds: ISO8601DateFormatter

    init(tailByteLimit: UInt64 = 512 * 1024) {
        self.tailByteLimit = tailByteLimit
        self.isoWithFractionalSeconds = ISO8601DateFormatter()
        self.isoWithFractionalSeconds.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        self.isoWithoutFractionalSeconds = ISO8601DateFormatter()
        self.isoWithoutFractionalSeconds.formatOptions = [.withInternetDateTime]
    }

    func parse(url: URL, modifiedAt: Date) throws -> SessionActivity? {
        let text = try tailText(from: url)
        var activity = SessionActivity(fileURL: url, modifiedAt: modifiedAt)

        for line in text.split(separator: "\n") {
            guard let event = parseEvent(from: String(line)) else {
                continue
            }

            activity.sawAnyEvent = true
            activity.latestEventAt = event.timestamp
            activity.latestEventType = event.eventType
            if activity.title == nil {
                activity.title = event.titleCandidate
            }
            if let contentText = event.contentText {
                activity.lastAnswer = contentText
            }

            if event.topLevelType == "event_msg" {
                switch event.eventType {
                case "task_started":
                    activity.lastTaskStartedAt = event.timestamp
                case "task_complete":
                    activity.lastTaskCompletedAt = event.timestamp
                default:
                    break
                }
            }
        }

        return activity.sawAnyEvent ? activity : nil
    }

    private func tailText(from url: URL) throws -> String {
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }

        let fileSize = try handle.seekToEnd()
        let bytesToRead = min(fileSize, tailByteLimit)
        let offset = fileSize - bytesToRead
        try handle.seek(toOffset: offset)

        let data = try handle.readToEnd() ?? Data()
        var text = String(decoding: data, as: UTF8.self)

        if offset > 0, let firstNewline = text.firstIndex(of: "\n") {
            text.removeSubrange(text.startIndex...firstNewline)
        }

        return text
    }

    private func parseEvent(from line: String) -> ParsedSessionEvent? {
        guard let data = line.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let timestampText = object["timestamp"] as? String,
              let timestamp = parseTimestamp(timestampText),
              let topLevelType = object["type"] as? String
        else {
            return nil
        }

        let payload = object["payload"] as? [String: Any]
        let eventType = payload?["type"] as? String ?? topLevelType
        let titleCandidate = SessionTitleNormalizer.explicitTitle(in: object)
            ?? userTitleCandidate(from: payload)
        let contentText = extractContentText(topLevelType: topLevelType, payload: payload, object: object)
        return ParsedSessionEvent(
            timestamp: timestamp,
            topLevelType: topLevelType,
            eventType: eventType,
            titleCandidate: titleCandidate,
            contentText: contentText
        )
    }

    private func extractContentText(topLevelType: String, payload: [String: Any]?, object: [String: Any]) -> String? {
        switch topLevelType {
        case "assistant":
            break
        default:
            return nil
        }

        let message = object["message"] as? [String: Any]
        return SessionTitleNormalizer.title(fromContent: message?["content"] ?? payload?["content"], maxLength: 200)
    }

    private func parseTimestamp(_ timestamp: String) -> Date? {
        isoWithFractionalSeconds.date(from: timestamp) ?? isoWithoutFractionalSeconds.date(from: timestamp)
    }

    private func userTitleCandidate(from payload: [String: Any]?) -> String? {
        guard payload?["role"] as? String == "user" else {
            return nil
        }

        return SessionTitleNormalizer.title(fromContent: payload?["content"])
    }
}

private struct ParsedSessionEvent {
    let timestamp: Date
    let topLevelType: String
    let eventType: String
    let titleCandidate: String?
    let contentText: String?
}

File: win/OneMenu.Core/Monitors/CodexStatusMonitor.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using OneMenu.Core.Infrastructure;
using OneMenu.Core.Models;

namespace OneMenu.Core.Monitors;

public partial class CodexStatusMonitor
{
    private readonly string _codexHome;
    private readonly TimeSpan _staleAfter;
    private readonly string _sessionsDirectory;
    private readonly string _titleIndexPath;
    private readonly int _scanLimit;
    private readonly JsonlFileReader _reader;
    private readonly JsonParser _parser;

    public string CodexHome => _codexHome;

    public CodexStatusMonitor(
        string? codexHome = null,
        TimeSpan? staleAfter = null,
        int scanLimit = 100)
    {
        _codexHome = codexHome
            ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".codex");
        _staleAfter = staleAfter ?? TimeSpan.FromMinutes(30);
        _scanLimit = scanLimit;
        _sessionsDirectory = Path.Combine(_codexHome, "sessions");
        _titleIndexPath = Path.Combine(_codexHome, "session_index.jsonl");
        _reader = new JsonlFileReader();
        _parser = new JsonParser();
    }

    public CodexStatusSnapshot Snapshot(DateTime? now = null)
    {
        var nowValue = now ?? DateTime.UtcNow;

        if (!Directory.Exists(_sessionsDirectory))
        {
            return CodexStatusSnapshot.Empty(_staleAfter, _codexHome, "没有找到 Codex sessions 目录");
        }

        try
        {
            var files = RecentSessionFiles(_scanLimit);
            var titleIndex = LoadTitleIndex();
            SessionActivity? latestActivity = null;
            var activeActivities = new List<SessionActivity>();
            var idleActivities = new List<SessionActivity>();

            foreach (var file in files)
            {
                var activity = _parser.Parse(file.path, file.modifiedAt);
                if (activity == null) continue;

                if (TryExtractSessionId(file.path, out var sessionId))
                {
                    if (titleIndex.TryGetValue(sessionId, out var indexedTitle))
                        activity.Title = indexedTitle;
                }

                if (IsNewer(activity, latestActivity))
                    latestActivity = activity;

                if (activity.IsOpenTask && IsFresh(activity, nowValue))
                    activeActivities.Add(activity);
                else
                    idleActivities.Add(activity);
            }

            var representative = NewestActivity(activeActivities) ?? latestActivity;
            var activeSessions = SessionSummaries(activeActivities);
            var idleSessions = SessionSummaries(idleActivities);

            return new CodexStatusSnapshot(
                State: activeActivities.Count == 0 ? CodexState.Idle : CodexState.Thinking,
                ActiveSessionCount: activeActivities.Count,
                LatestEventAt: representative?.LatestEventAt,
                LatestEventType: representative?.LatestEventType,
                LatestSessionFile: representative?.FilePath,
                LatestSessionTitle: representative != null ? SessionTitle(representative) : null,
                ActiveSessions: activeSessions,
                IdleSessions: idleSessions,
                ActiveSessionTitles: activeSessions.Select(s => s.Title).ToArray(),
                IdleSessionTitles: idleSessions.Select(s => s.Title).ToArray(),
                ScannedFileCount: files.Count,
                StaleAfter: _staleAfter,
                CodexHome: _codexHome,
                ErrorMessage: null);
        }
        catch (Exception ex)
        {
            return CodexStatusSnapshot.Empty(_staleAfter, _codexHome, ex.Message);
        }
    }

    private List<(string path, DateTime modifiedAt)> RecentSessionFiles(int limit)
    {
        var files = new List<(string path, DateTime modifiedAt)>();

        try
        {
            foreach (var filePath in Directory.EnumerateFiles(_sessionsDirectory, "*.jsonl",
                SearchOption.AllDirectories))
            {
                try
                {
                    var fileInfo = new FileInfo(filePath);
                    files.Add((filePath, fileInfo.LastWriteTimeUtc));
                }
                catch
                {
                    // skip inaccessible files
                }
            }
        }
        catch
        {
            // directory enumeration failed
        }

        return files
            .OrderByDescending(f => f.modifiedAt)
            .Take(limit)
            .ToList();
    }

    private Dictionary<string, string> LoadTitleIndex()
    {
        var result = new Dictionary<string, string>();
        if (!File.Exists(_titleIndexPath)) return result;

        try
        {
            foreach (var line in File.ReadLines(_titleIndexPath))
            {
                if (string.IsNullOrWhiteSpace(line)) continue;
                try
                {
                    using var doc = JsonDocument.Parse(line);
                    var root = doc.RootElement;
                    if (root.TryGetProperty("id", out var idEl) &&
                        idEl.ValueKind == JsonValueKind.String)
                    {
                        var title = SessionTitleNormalizer.ExplicitTitleIn(root);
                        if (title != null)
                            result[idEl.GetString()!] = title;
                    }
                }
                catch
                {
                    // skip malformed lines
                }
            }
        }
        catch
        {
            // file read failed
        }

        return result;
    }

    private bool IsFresh(SessionActivity activity, DateTime now)
    {
        var freshnessDate = Max(activity.LatestEventAt ?? activity.ModifiedAt, activity.ModifiedAt);
        return (now - freshnessDate) <= _staleAfter;
    }

    private static bool IsNewer(SessionActivity activity, SessionActivity? other)
    {
        if (other == null) return true;
        return (activity.LatestEventAt ?? activity.ModifiedAt) >
               (other.LatestEventAt ?? other.ModifiedAt);
    }

    private static SessionActivity? NewestActivity(List<SessionActivity> activities)
    {
        return activities.MaxBy(a => a.LatestEventAt ?? a.ModifiedAt);
    }

    private static List<StatusSessionSummary> SessionSummaries(List<SessionActivity> activities)
    {
        return activities
            .OrderByDescending(a => a.LatestEventAt ?? a.ModifiedAt)
            .Select(a => new StatusSessionSummary(
                Id: SessionIdentifierFor(a),
                Title: SessionTitle(a),
                LastAnswer: a.LastAnswer))
            .ToList();
    }

    private static string SessionTitle(SessionActivity activity) =>
        SessionTitleNormalizer.DisplayTitle(activity.Title);

    private static string SessionIdentifierFor(SessionActivity activity) =>
        TryExtractSessionId(activity.FilePath, out var id) ? id : activity.FilePath;

    private static bool TryExtractSessionId(string filePath, out string sessionId)
    {
        sessionId = "";
        var baseName = Path.GetFileNameWithoutExtension(filePath);
        if (baseName.Length < 36) return false;

        var suffix = baseName[^36..];
        if (UuidRegex().IsMatch(suffix))
        {
            sessionId = suffix;
            return true;
        }
        return false;
    }

    private static DateTime Max(DateTime a, DateTime b) => a > b ? a : b;

    [GeneratedRegex(@"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$")]
    private static partial Regex UuidRegex();

    /// <summary>
    /// JSONL event parser for Codex session files.
    /// </summary>
    private class JsonParser
    {
        private static readonly string[] IsoFormats =
        {
            "yyyy-MM-dd'T'HH:mm:ss.FFFFFFF'Z'",
            "yyyy-MM-dd'T'HH:mm:ss.FFFFFFFzzz",
            "yyyy-MM-dd'T'HH:mm:ss'Z'",
            "yyyy-MM-dd'T'HH:mm:sszzz"
        };

        public SessionActivity? Parse(string filePath, DateTime modifiedAt)
        {
            try
            {
                var text = new JsonlFileReader().ReadTail(filePath);
                var activity = new SessionActivity
                {
                    FilePath = filePath,
                    ModifiedAt = modifiedAt
                };

                foreach (var line in text.Split('\n'))
                {
                    var trimmed = line.Trim();
                    if (trimmed.Length == 0) continue;

                    var evt = ParseEvent(trimmed);
                    if (evt == null) continue;

                    activity.SawAnyEvent = true;
                    activity.LatestEventAt = evt.Value.Timestamp;
                    activity.LatestEventType = evt.Value.EventType;
                    activity.Title ??= evt.Value.TitleCandidate;
                    if (evt.Value.ContentText != null)
                        activity.LastAnswer = evt.Value.ContentText;

                    if (evt.Value.TopLevelType == "event_msg")
                    {
                        switch (evt.Value.EventType)
                        {
                            case "task_started":
                                activity.LastTaskStartedAt = evt.Value.Timestamp;
                                break;
                            case "task_complete":
                                activity.LastTaskCompletedAt = evt.Value.Timestamp;
                                break;
                        }
                    }
                }

                return activity.SawAnyEvent ? activity : null;
            }
            catch
            {
                return null;
            }
        }

        private (DateTime Timestamp, string TopLevelType, string EventType, string? TitleCandidate,
            string? ContentText)? ParseEvent(string line)
        {
            try
            {
                using var doc = JsonDocument.Parse(line);
                var root = doc.RootElement;

                if (!root.TryGetProperty("timestamp", out var tsEl) ||
                    tsEl.ValueKind != JsonValueKind.String)
                    return null;
                if (!TryParseTimestamp(tsEl.GetString()!, out var timestamp))
                    return null;
                if (!root.TryGetProperty("type", out var typeEl) ||
                    typeEl.ValueKind != JsonValueKind.String)
                    return null;

                var topLevelType = typeEl.GetString()!;
                JsonElement? payload = root.TryGetProperty("payload", out var pl) &&
                    pl.ValueKind == JsonValueKind.Object ? pl : null;
                var eventType = (payload?.TryGetProperty("type", out var et) == true &&
                    et.ValueKind == JsonValueKind.String) ? et.GetString()! : topLevelType;

                var titleCandidate = SessionTitleNormalizer.ExplicitTitleIn(root)
                    ?? UserTitleCandidate(payload);

                string? contentText = ExtractContentText(topLevelType, payload, root);

                return (timestamp, topLevelType, eventType, titleCandidate, contentText);
            }
            catch
            {
                return null;
            }
        }

        private static string? ExtractContentText(string topLevelType, JsonElement? payload, JsonElement root)
        {
            if (topLevelType != "assistant")
                return null;

            JsonElement? message = root.TryGetProperty("message", out var msg) &&
                msg.ValueKind == JsonValueKind.Object ? msg : null;

            JsonElement? content = message?.TryGetProperty("content", out var mc) == true
                ? mc : payload?.TryGetProperty("content", out var pc) == true
                    ? pc : null;

            if (content == null) return null;
            return SessionTitleNormalizer.TitleFromContent(content.Value, 200);
        }

        private static string? UserTitleCandidate(JsonElement? payload)
        {
            if (payload == null) return null;
            if (payload.Value.TryGetProperty("role", out var role) &&
                role.ValueKind == JsonValueKind.String &&
                role.GetString() == "user")
            {
                if (payload.Value.TryGetProperty("content", out var content))
                    return SessionTitleNormalizer.TitleFromContent(content);
            }
            return null;
        }

        private static bool TryParseTimestamp(string text, out DateTime result)
        {
            // ISO 8601 with optional fractional seconds and timezone
            if (DateTime.TryParseExact(text, IsoFormats,
                    System.Globalization.CultureInfo.InvariantCulture,
                    System.Globalization.DateTimeStyles.AssumeUniversal |
                    System.Globalization.DateTimeStyles.AdjustToUniversal,
                    out result))
                return true;

            // Fallback to standard .NET ISO parser
            if (DateTime.TryParse(text, null,
                    System.Globalization.DateTimeStyles.RoundtripKind, out result))
            {
                result = result.ToUniversalTime();
                return true;
            }

            return false;
        }
    }
}
